### PR TITLE
Scintilla 3.21.1

### PR DESCRIPTION
**Pull request by MaartenBent**
Demonstrate updating Scintilla to version ~3.10.5~ ~3.11.0~ ~3.20.0~ ~3.21.0~ 3.21.1 of the long-term branch (https://scintilla.sourceforge.io/LongTermDownload.html).

Scintilla requires a c++11 compiler, so currently it cannot be merged into wxWidgets master.

Scintilla is an external library that can/should be included as a submodule instead of copying all the sources manually.

Also see #668, which proposes to use two scintilla versions side-by-side.

### CODE REVIEW

**Comment by pkulchenko** [comment]
I don't have a problem with having two Scintilla versions side-by-side, especially given that the existing version of Scintilla we bundle with wxwidgets is a dead end. We probably should *not* pick the newer one automatically (even when C++11 compiler is present), but I think it should be available as an option and should generate the same library.

**Comment by vadz** [comment]
Maarten, am I supposed to do anything with this one or is it just an experiment?

**Comment by MaartenBent** [comment]
For now there is not much you can do. Maybe add a 3.3 milestone or appropriate label.

**Comment by vadz** [comment]
OK, thanks, please clear the "postponed" label when it's time to integrate this.

**Comment by imciner2** [comment]
Having to use the old scintilla version (3.7.x) is getting somewhat painful, because we just hit a bug related to IME input that is fixed in upstream Scintilla apparently (https://gitlab.com/kicad/code/kicad/-/issues/6229#note_525462051). It was 2 years almost since this PR was postponed to 3.3, and 3.2 is still not out the door so saying it won't be updated until 3.3 seems to be saying we will need to wait a few more years before we can have the bugs in the Scintilla engine fixed.

**Comment by vadz** [comment]
The underlying problem is that we can't require C++11 in wx 3.2 because the jump would be too big (we'd go from supporting MSVC 6 in 3.0 to requiring MSVS 2013 or maybe even 2015 in 3.2) and we don't even use C++11 internally. I'd have no qualms whatsoever to require C++14 or maybe even later in 3.4, but we can't release 3.4 without releasing 3.2 first, can we...

So I wonder if it could be acceptable to create 3.3 branch and start using C++14 on it already. Then the new Scintilla version could be merged there right now. Of course, this doesn't help at all if you want to use a released wx version. But OTOH refactoring Scintilla support it's inevitably going to move the 3.2 release back a bit more.

But yes, if we have to include newer Scintilla in 3.2 Paul's suggestion is probably the only workable one. I still don't like it because it's awfully surprising to get a different library version (and hence different functionality and different bugs) just due to a change of `-std=xxx` option in your makefile. But I can't propose anything else.

**Comment by NewPagodi** [comment]
A possible option if someone needs a newer Scintilla version right now is to use wxNativeWindow to wrap a Scintilla window/widget/view.  This is what I do when I want to compare the wxSTC implementation of a feature with the implementation provided by the Scintilla developers.

**Comment by pkulchenko** [comment]
> But yes, if we have to include newer Scintilla in 3.2 Paul's suggestion is probably the only workable one. I still don't like it because it's awfully surprising to get a different library version (and hence different functionality and different bugs) just due to a change of -std=xxx option in your makefile. But I can't propose anything else.

Just to be clear, I'm not (necessarily) suggesting that the switch should happen automatically. I suggest we do the following: (1) package the long term version, as this proposed patch does and (2) have STCLONGTERM make option that enables usage of this branch and fails if its C++ version condition is not satisfied. This way even those users that have a newer version of C++ compiler will get the same Scintilla version unless the use STCLONGTERM option and we still get a way to upgrade Scintilla.

My recommendation is to do this (add the Scintilla long-term source) as quickly as possible, as it shouldn't affect anything else in wxiwdgets (assuming it's only enabled with the STCLONGTERM option, and it is off by default).

**Comment by MaartenBent** [comment]
I could update this PR to make this work with CMake, but I don't know if and how this can be done in the bakefile-generated makefiles / project files or the vcxproj files.
(Or probably a different PR, it could be useful to keep this one around. I can at least rebase this PR and update scintilla to the latest version again.)

Some todos that come to mind:
- Create a stc3 source directory with the updated stc files (from this PR) and scintilla 3.21 sources (could be in a submodule as well).
- In CMake, create the wxstc library from either the 3.7 (stc) or the 3.21 (stc3) sources based on `STCLONGTERM` (working title) build option.
- In bakefile and vcxproj, investigate how to do this.
- `stc.h` (include and interface) should now be generated from the stc3 sources. All new functions should be wrapped with something like `#ifdef STCLONGTERM`. Or a dummy implementation of these functions in the 3.7 sources?
- Remove the now unused .py and .in files from the stc directory.

Other considerations:
- Can it be extended to Scintilla 4 (c++14 ?) or Scintilla 5 (C++17)?
- The developer including `stc.h` will need to define `STCLONGTERM` to use the new functions. Is it in `wx-config --cxxflags`? Or in `setup.h`? Maybe a better name is `wxUSE_SCINTILLA_CXX11`.
- Enable `STCLONGTERM` in the build scripts for the official releases? Will linux package maintainers enable this?  Otherwise only people who build and distribute their own version of wxWidgets can benefit from this.
- Or always enable it when building with c++11 / vc140 and add a build option to disable it?

Are there any other obvious things that I missed?

**Comment by imciner2** [comment]
Adding it in alongside the current one would work for my use case - since I basically want to have the updated version available (and we ship custom compiled versions on macOS and FlatPak, so enabling a flag to select this is fine).

> The developer including stc.h will need to define STCLONGTERM to use the new functions. Is it in wx-config --cxxflags? Or in setup.h? Maybe a better name is wxUSE_SCINTILLA_CXX11.

I think it would be nicer to have it inside `setup.h` and then have that included from `stc.h` automatically.

> Enable STCLONGTERM in the build scripts for the official releases? Will linux package maintainers enable this? Otherwise only people who build and distribute their own version of wxWidgets can benefit from this.

It is up to the individual packagers, but I think they will probably enable it when requested if their compilers support it.

> Or always enable it when building with c++11 / vc140 and add a build option to disable it?

I'd be for this change, but it might be difficult to do this because we would need to figure out a way of identifying all the compilers and versions it would need to be disabled for.

**Comment by vadz** [comment]
Sorry, I didn't have time at all to look at/think about this, but just some random thoughts:

- It would be better to use a submodule rather than just a subdirectory, this should hopefully make tracking upstream easier.
- If it can't be made to work with bkl, could we just make it work in (manually maintained) `.vcxproj` files and configure? We don't care about `.vcproj` ones anyhow, they are used with compilers without C++11 support. `makefile.gcc` might be more relevant, but if it at least works with configure, it's not too bad.
- As long as we can test compiler C++ support (possible with CMake, configure and vcxproj files with MSBuild conditions), we could automatically select the latest and greatest version, even Scintilla 5. I don't know if we need to provide manual override, i.e. if there are any reasons to want to use an older version.

Thanks for looking at this!

**Comment by pkulchenko** [comment]
> As long as we can test compiler C++ support (possible with CMake, configure and vcxproj files with MSBuild conditions), we could automatically select the latest and greatest version, even Scintilla 5.

I'd personally prefer to have an option, as there are some incompatibilities between Scintilla versions (mostly related to Lexilla and Scintillua handling as far as I understand).

>  I don't know if we need to provide manual override, i.e. if there are any reasons to want to use an older version.

I think auto-selection of the latest is fine, but I should be able to pick an older version if I need.

**Comment by vadz** [comment]
AFAICS we should be able to use this version of Scintilla with C++14 only, but latest versions require C++17. We need to decide if support for the latest version is important enough for us and, if so, whether we can use 2 Scintilla versions in parallel and select the one compatible with the current compiler, or just require C++17 compiler for building wx itself too.

**Comment by MaartenBent** [comment]
The Scintilla long-term branch doesn't seem to get updates anymore. So If we want to include a more recent version, it requires c++17. But seeing how Scintilla adopts new versions, this might become c++20 before wx3.3.

A while back I looked at having two versions in parallel, and I had it working with CMake. But with bakefile or MSVC projects it was not possible (for me), so I never finished it.
I had the current version and the one from this PR in parallel, and based on `CMAKE_CXX_STANDARD`, or an explicit option, a version was selected.
The same can be done in wx3.3 to choose between this version (or the last version before c++17 was required), and a version with the latest release (c++17 or maybe even c++20).

Another options is to require c++14 for wxWidgets, but require c++17 when `wxUSE_STC` is enabled.

**Comment by vadz** [comment]
> A while back I looked at having two versions in parallel, and I had it working with CMake. But with bakefile or MSVC projects it was not possible (for me), so I never finished it. I had the current version and the one from this PR in parallel, and based on `CMAKE_CXX_STANDARD`, or an explicit option, a version was selected. The same can be done in wx3.3 to choose between this version (or the last version before c++17 was required), and a version with the latest release (c++17 or maybe even c++20).

We probably should do this, I should be able to implement the same thing with configure. For MSVS projects we just have to make the Scintilla version conditional on MSVS version, so it should be doable too. And we won't be using bakefile anyhow soon -- even though it's not clear what will replace it.

> Another options is to require c++14 for wxWidgets, but require c++17 when `wxUSE_STC` is enabled.

If we keep `makefile.gcc`, we ought to be able to add some `STC_VERSION` option and tell people to set it to 4 or 5 depending on whether they use new enough compiler. Or we could simply rely on people using `makefile.gcc` using recent enough gcc because this seems to be the case in practice: old gcc are only used on old (typically RHEL) Linux systems.

**Comment by MaartenBent** [comment]
This can be merged now.

I'll start investigating updating to 4 and/or 5.

**Comment by vadz** [comment]
Great, thanks!

File: build/bakefiles/scintilla.bkl
Comment (vadz) [inline build/bakefiles/scintilla.bkl:57]: Out of curiosity, why do we disable this one?

File: build/osx/wxiphone.xcodeproj/xcshareddata/xcschemes/static.xcscheme
Comment (vadz) [inline build/osx/wxiphone.xcodeproj/xcshareddata/xcschemes/static.xcscheme:3]: Is this intentional? Important? @csomor It it ok to merge this?

**Review by vadz — APPROVED** [review]
Thanks for all your work on this, I obviously didn't really look at all the changes, but what I saw looks good and, in any case, I think we need to merge it to get some testing as any problems almost surely won't be discovered by just reading the code. Of course, if more people can test this before the merge, it would be very great too, but otherwise I'll merge it -- unless you decide to change the explicit Scintilla additions as I think this adds a lot of "noise" to the patch and might be unnecessary.

I also have another question: should we move `src/stc/scintilla` contents in a submodule? I think it might make it easier to update it in the future and it definitely makes it simpler to avoid making global changes to the files there.

Thanks again!

File: build/bakefiles/scintilla.bkl
Comment (MaartenBent) [reply build/bakefiles/scintilla.bkl:57]: When enabled, it needs external lua files that are not part of the scintilla package:
https://github.com/wxWidgets/wxWidgets/blob/06dbd32194b2978b6162d61cbe6237eea7b1b661/src/stc/scintilla/lexers/LexLPeg.cxx#L39-L44

File: build/osx/wxiphone.xcodeproj/xcshareddata/xcschemes/static.xcscheme
Comment (MaartenBent) [reply build/osx/wxiphone.xcodeproj/xcshareddata/xcschemes/static.xcscheme:3]: I just ran the python script. Maybe different Xcode version? Mine is 13.2.

File: src/stc/PlatWX.cpp
Comment (MaartenBent) [reply src/stc/PlatWX.cpp:3453]: I'll see if I can add that.

**Comment by MaartenBent** [comment]
A submodule would be nice and should eventually be used, but it has some problems. I don't think there is any up-to-date git upstream we can use. And with Scintilla 5, the project has split into two parts: Scintilla (using mercurial on sourceforge) and Lexilla on github.
And then we have to decide if and how we can support multiple versions of Scintilla for different c++ versions. Put them all in one submodule in different folders, or different submodules for each.

File: build/osx/wxiphone.xcodeproj/xcshareddata/xcschemes/static.xcscheme
Comment (MaartenBent) [reply build/osx/wxiphone.xcodeproj/xcshareddata/xcschemes/static.xcscheme:3]: It also changed file order and many identifiers in the pbxproj files. I don't know if this is expected or if it is possible to only add/remove the changed scintilla files.

File: build/bakefiles/scintilla.bkl
Comment (pkulchenko) [reply build/bakefiles/scintilla.bkl:57]: FYI, lpeg lexer can be built and loaded separately, so no functionality is really lost if it's not included by default (although it's not going to be statically linked). It uses `LoadLexerLibrary` and it does require a bit of a setup (for example, `lexer.lpeg.home` property may need to be set). I've been using this lexer/approach for many years in my wxiwdgets/wxlua-based project and can provide some additional details if needed.

Another option is to bundle those Lua files, but unfortunately they are (Lua) version dependent and will load a Lua library at run-time, so not going to work without it being present anyway.

**Comment by vadz** [comment]
Thanks again for your work on this, I'm merging and pushing it and we can discuss switching submodules later if you think it's worth it.

**Pull request merged**
